scan-external-contacts/nodejs/index.js

```diff
@@ -0,0 +1,325 @@
+
+        // Open write stream for file output
+        let output_filename, headerString, footerString;
+        let firstWrite = true;
+
+        if (exportType === 'CSV') {
+            console.log("Export to CSV");
+            output_filename = "GenesysCloud_ExternalContacts_Export_" + (Date.now()).toString() + ".csv";
+            // We will export to a csv format - so we will need a first line with the column/header names we want to export
+            headerString = "id" + fieldSeparator + "modifyDate" + fieldSeparator + "firstName" + fieldSeparator + "lastName" + fieldSeparator + "workPhone" + fieldSeparator + "workEmail" + fieldSeparator + "externalOrganizationId" + fieldSeparator + "surveyOptOut" + "\n";
+        } else {
+            console.log("Export to Raw JSON");
+            output_filename = "GenesysCloud_ExternalContacts_Export_" + (Date.now()).toString() + ".json";
+            // We will export to a raw json format (JSON Array) - so we will need to open the JSON array
+            headerString = "[";
+        }
+
+        writeStream = fs.createWriteStream(output_filename);
+
+        // Write Header for Contacts (if necessary)
+        if (writeStream) {
+            if (headerString) {
+                writeStream.write(headerString);
+            }
+        }
+
+        // Starting API scan of External Contacts
+        let isCompleted = false;
+        // First scan - cursor = null
+        let currentCursor = null;
+
+        while (isCompleted === false) {
+            console.log("Scan External Contacts - cursor: ", currentCursor);
+            let scanResult = await gcloudContactsScan(currentCursor);
+
+            if (scanResult.error === false) {
+                if (scanResult.isEmpty === false) {
+                    // Process current data (filter, transform, write)
+                    // Note: filter - You can implement a filter condition here and decide if the entity is to be printed (proceed) or not (return)
+                    // Note: transform - You can manipulate the entity attributes to create the desired string to be printed
+
+                    if (exportType === 'CSV') {
+                        // CSV
+                        if (writeStream) {
+                            for (let entity of scanResult.entities) {
+                                // For each entity element, print a line (CSV) with a field separator between values
+                                let entityString = (entity.id ? entity.id : '') + fieldSeparator + (entity.modifyDate ? entity.modifyDate : '') + fieldSeparator + (entity.firstName ? entity.firstName : '') + fieldSeparator + (entity.lastName ? entity.lastName : '') + fieldSeparator + (entity.workPhone ? entity.workPhone.e164 : '') + fieldSeparator + (entity.workEmail ? entity.workEmail : '') + fieldSeparator + (entity.externalOrganization ? entity.externalOrganization.id : '') + fieldSeparator + (entity.surveyOptOut ? entity.surveyOptOut.toString() : 'false') + "\n";
```

Comment:
Don't build CSV files manually. Use a [node library for CSV handling](https://www.npmjs.com/search?q=csv) to ensure proper field escaping and file formatting.